2.5-huffman.ipynb

```diff
@@ -1,25 +1,248 @@
 {
  "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import util\n",
+    "from test_framework.key import ECKey, ECPubKey, generate_key_pair\n",
+    "from test_framework.script import TapTree, TapLeaf, Node"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# 2.5 Huffman TapTree Constructor\n",
+    "\n",
+    "When constructing a TapTree, there is the possibility of optimizing the tree structure so that the expected witness weight when spending along the script path is minimized. The `expected witness weight` is determined by summing up all probability-adjusted witness weights of the TapScripts in the TapTree.\n",
+    "\n",
+    "For example, for a TapTree with TapLeafs A, B, and C, the expected witness weight is determined as follows:\n",
+    "\n",
+    "* **`Expected-witness-weight =`**\n",
+    "    * **`Probability-of-A * Witness-weight-A`**\n",
+    "    * **`+ Probability-of-B * Witness-weight-B`**\n",
+    "    * **`+ Probability-of-C * Witness-weight-C`**\n",
+    "\n",
+    "In order to minimize the expected witness weight for the script path, we can try to reduce the size of the required `inclusion proof` for those TapScripts which have a higher probability weighting by placing these closer to the root of the tree."
```

Comment:
I think it's clearer if you replace 'probability weighting' with 'expected weight' here and throughout the chapter.

Reply:
Thanks John, however, I am not sure I understand why it is clearer.

* The overall expected witness weight is the sum of witness weights * probability (based on probability distribution).
* The probability weighting captures the relative probability for a given leaf.

An "expected" value should be adjusted for probability-distribution, right? 
Shouldn't we reserve the term "expected" for this?

Reply:
Hmmm, yes perhaps you're right that _expected_ should only be used for the distribution. I think the most accurate term we're looking for is _probability-weighted weight_, but that sounds a bit clumsy.

I've just realised that you're assuming that the tapscript witness is the same weight for all tapscripts in this tree, and I think you're using the word 'weight' lower down in this notebook for 'probability'. Is that correct? If so, I think it's pretty confusing.